src/utils/translator.rs

```diff
@@ -0,0 +1,494 @@
+	}
+}
+
+pub fn refty(t: &mut dyn Translator, ty: &wasmparser::RefType) -> Result<RefType> {
+	Ok(RefType { nullable: ty.is_nullable(), heap_type: t.translate_heapty(&ty.heap_type())? })
+}
+
+pub fn heapty(t: &mut dyn Translator, ty: &wasmparser::HeapType) -> Result<HeapType> {
+	match ty {
+		wasmparser::HeapType::Func => Ok(HeapType::Func),
+		wasmparser::HeapType::Extern => Ok(HeapType::Extern),
+		wasmparser::HeapType::Any => Ok(HeapType::Any),
+		wasmparser::HeapType::None => Ok(HeapType::None),
+		wasmparser::HeapType::NoExtern => Ok(HeapType::NoExtern),
+		wasmparser::HeapType::NoFunc => Ok(HeapType::NoFunc),
+		wasmparser::HeapType::Eq => Ok(HeapType::Eq),
+		wasmparser::HeapType::Struct => Ok(HeapType::Struct),
+		wasmparser::HeapType::Array => Ok(HeapType::Array),
+		wasmparser::HeapType::I31 => Ok(HeapType::I31),
+		wasmparser::HeapType::Indexed(i) => Ok(HeapType::Indexed(t.remap(Item::Type, *i)?)),
+	}
+}
+
+pub fn global(t: &mut dyn Translator, global: Global, s: &mut GlobalSection) -> Result<()> {
+	let ty = t.translate_global_type(&global.ty)?;
+	let insn =
+		t.translate_const_expr(&global.init_expr, &global.ty.content_type, ConstExprKind::Global)?;
+	s.global(ty, &insn);
+	Ok(())
+}
+
+pub fn export_kind(_: &dyn Translator, kind: ExternalKind) -> Result<ExportKind> {
+	match kind {
+		ExternalKind::Table => Ok(ExportKind::Table),
+		ExternalKind::Global => Ok(ExportKind::Global),
+		ExternalKind::Tag => Ok(ExportKind::Tag),
+		ExternalKind::Func => Ok(ExportKind::Func),
+		ExternalKind::Memory => Ok(ExportKind::Memory),
+	}
+}
+
+#[allow(unused)]
+pub fn export(
+	t: &mut dyn Translator,
+	e: &wasmparser::Export<'_>,
+	sec: &mut wasm_encoder::ExportSection,
+) -> Result<()> {
+	sec.export(e.name, t.translate_export_kind(e.kind)?, e.index);
+	Ok(())
+}
+
+pub fn const_expr(
+	t: &mut dyn Translator,
+	e: &wasmparser::ConstExpr<'_>,
+	ctx: ConstExprKind,
+) -> Result<wasm_encoder::ConstExpr> {
+	let mut e = e.get_operators_reader();
+	let mut offset_bytes = Vec::new();
+	let op = e.read()?;
+	if let ConstExprKind::ElementFunction = ctx {
+		match op {
+			Operator::RefFunc { .. } |
+			Operator::RefNull { hty: wasmparser::HeapType::Func, .. } |
+			Operator::GlobalGet { .. } => {},
+			_ => return Err(TranslatorError::NoMutationsApplicable),
+		}
+	}
+	t.translate_op(&op)?.encode(&mut offset_bytes);
+	match e.read()? {
+		Operator::End if e.eof() => {},
+		_ => return Err(TranslatorError::NoMutationsApplicable),
+	}
+	Ok(wasm_encoder::ConstExpr::raw(offset_bytes))
+}
+
+#[allow(dead_code)]
+pub fn element(
+	t: &mut dyn Translator,
+	element: wasmparser::Element<'_>,
+	s: &mut ElementSection,
+) -> Result<()> {
+	let offset;
+	let mode = match &element.kind {
+		ElementKind::Active { table_index, offset_expr } => {
+			offset = t.translate_const_expr(
+				offset_expr,
+				&wasmparser::ValType::I32,
+				ConstExprKind::ElementOffset,
+			)?;
+			let table_index = table_index.unwrap_or(0);
```

Comment:
Why default to `0`?   We only need to map `Option::Some`?

If table_index is indeed `0`, you are setting the table to `None`.

Reply:
True. All of this is not needed, since table_index is an Option from the very beginning.